### PR TITLE
Optimize and minimize bound checks

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We should minimize and optimize bound checks as much as possible to get the most out of performance.

Modifications:

- Use bitwise operations to remove branching
- Remove branches when possible

Result:

Better performance for various operations.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@nmittler @Scottmitch please check!

File: buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java [hunk 1]
```diff
@@ -1145,22 +1149,30 @@ protected final void checkSrcIndex(int index, int length, int srcIndex, int srcC
...
         checkIndex(index, length);
-        if (dstIndex < 0 || dstIndex > dstCapacity - length) {
+        if (isInvalid(dstIndex, length, dstCapacity)) {
             throw new IndexOutOfBoundsException(String.format(
                     "dstIndex: %d, length: %d (expected: range(0, %d))", dstIndex, length, dstCapacity));
         }
     }
 
+    private static boolean isInvalid(int index, int length, int capacity) {
+        return (index | length | (index + length) | (capacity - (index + length))) < 0;
```
Comment (nmittler) [inline buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java]: clever :)

**Comment by nmittler** [comment]
@normanmaurer LGTM

File: buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java (hunk 1 above)
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java]: :heart_eyes: ;)

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (af9dc2c6a6c00eed4e93ae948349ad41c9b217f1), 4.1 (ffe7aafd8252a382aa0a651d3e9fb84065304da6) and master (b7b4277553f2468685ff64e818a5bd061997ded8)

**Comment by Scottmitch** [comment]
@normanmaurer - lgtm

**Pull request closed without merging**